### PR TITLE
Register and process Gravity PDF mergetag with GF Feeds

### LINKED ISSUE #960: Allow PDF Merge Tag to be used in AddOn Map Type
This will allow the PDF URL to be sent with API requests to other services.

The basis for this functionality is as follows:

```
add_filter( 'gform_addon_field_map_choices', function( $fields, $form_id, $field_type, $exclude_field_types ) {
	$pdfs = GPDFAPI::get_form_pdfs( $form_id );

	foreach ( $pdfs as $pdf ) {
		$merge_tag = sprintf(
			'{%s:pdf:%s}',
			$pdf['name'],
			$pdf['id']
		);

		$fields[] = [
			'label' => $merge_tag,
			'value' => $merge_tag,
		];
	}

	return $fields;
}, 10, 4 );

add_filter( 'gform_addon_field_value', function( $field_value, $form, $entry, $field_id ) {
	if ( ! empty( $field_value ) || strpos( $field_id, ':pdf:' ) === false ) {
		return $field_value;
	}

	$mergetags = GPDFAPI::get_mvc_class( 'Model_Mergetags' );

	return $mergetags->process_pdf_mergetags( $field_id, $form, $entry, false );
}, 10, 4 );
```

### PR DESCRIPTION
**Pull request by jakejackson1**
## Description
WIP

Resolves #960

## Checklist:
- [ ] I've tested the code.
- [ ] My code is easy to read, follow, and understand
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation / docblocks.

### CODE REVIEW

**Comment by jakejackson1** [comment]
@jestonihpi From my testing of this list of Gravity Forms plugins https://github.com/GravityPDF/gravity-pdf/issues/960#issuecomment-671752728 this PR will work provided user's don't select the "Other value" option and then try paste in the Gravity PDF merge tag. Processing merge tags in the "Other value" appears very inconsistent across each Gravity Forms add-on, and they look to be handled multiple different ways (for legacy reasons).

With the above in mind, let's forget about supporting the "Other value" option. This PR should instead be updated to also register 3 signed PDF URLs for each PDF:

* One Day
* One Month
* One Year

This will give people more flexibility in how they use the PDF URLs, and make it more secure.

File: src/Model/Model_Mergetags.php [hunk 1]
```diff
@@ -244,4 +244,63 @@ public function process_pdf_mergetags( $text, $form, $entry, $url_encode ) {
...
+	 * @param string $field_value
+	 * @param int $form_id
+	 * @param int $field_id
+	 * @param array $entry
+	 *
+	 * @return string
+	 *
+	 * @since 6.0
+	 */
+	public function legacy_addon_field_value( $field_value, $form_id, $field_id, $entry ) {
```
Comment (jakejackson1) [inline src/Model/Model_Mergetags.php]: delete this

File: src/Model/Model_Mergetags.php [hunk 2]
```diff
@@ -244,4 +244,63 @@ public function process_pdf_mergetags( $text, $form, $entry, $url_encode ) {
...
+	 * @return string
+	 *
+	 * @since 6.0
+	 */
+	public function process_field_value( $field_value, $form, $entry, $field_id ) {
+		if ( ! empty( $field_value ) || strpos( $field_id, ':pdf:' ) === false ) {
+			return $field_value;
+		}
+
+		return $this->process_pdf_mergetags( $field_id, $form, $entry, false );
```
Comment (jakejackson1) [inline src/Model/Model_Mergetags.php]: use $gform->process_tags()

File: src/Model/Model_Mergetags.php [hunk 3]
```diff
@@ -244,4 +244,63 @@ public function process_pdf_mergetags( $text, $form, $entry, $url_encode ) {
...
+	 */
+	public function add_field_map_choices( $fields, $form_id ) {
+		foreach ( $this->options->get_form_pdfs( $form_id ) as $pdf ) {
+			$merge_tag = sprintf(
+				'{%s:pdf:%s}',
+				$pdf['name'],
+				$pdf['id']
+			);
+
+			$fields[] = [
```
Comment (jakejackson1) [inline src/Model/Model_Mergetags.php:316]: register signed merge tag: 1 week, 1 month, 1 year

File: src/Controller/Controller_Mergetags.php [hunk 4]
```diff
@@ -61,5 +61,12 @@ public function init() {
 	public function add_filters() {
 		add_filter( 'gform_replace_merge_tags', [ $this->model, 'process_pdf_mergetags' ], 10, 4 );
 		add_filter( 'gform_custom_merge_tags', [ $this->model, 'add_pdf_mergetags' ], 10, 2 );
+
+		add_filter( 'gform_addon_field_map_choices', [ $this->model, 'add_field_map_choices' ], 10, 2 );
+		add_filter( 'gform_addon_field_value', [ $this->model, 'process_field_value' ], 10, 4 );
+
+		add_filter( 'gform_mailchimp_field_value', [ $this->model, 'legacy_addon_field_value' ], 10, 4 );
```
Comment (jakejackson1) [inline src/Controller/Controller_Mergetags.php]: Remove add-on specific filters

**Review by jakejackson1 — COMMENTED** [review]
Add unit tests

**Comment by jakejackson1** [comment]
@jestonihpi `$this->options->get_form_pdfs( $form_id )` can return a `WP_Error` which we should handle in `Model_Mergetags::add_field_map_choices()` before trying to loop over the results.

Also, which Gravity Forms add-ons did you test this code with to ensure the PDF URLs were successfully passed to the service?

**Comment by jestonihpi** [comment]
> Also, which Gravity Forms add-ons did you test this code with to ensure the PDF URLs were successfully passed to the service?

I only used the PostCreation Add-on on this development Jake.

**Comment by jestonihpi** [comment]
image

This option shouldn't exists if there are no PDF template found.

**Comment by jakejackson1** [comment]
> image
> 
> This option shouldn't exists if there are no PDF template found.



> image
> 
> This option shouldn't exists if there are no PDF template found.

Fixed in latest commit

**Comment by jakejackson1** [comment]
> image
> 
> This option shouldn't exists if there are no PDF template found.

Fixed in latest commit

**Pull request merged**